### PR TITLE
Move typetraits.`$` to system. Fixes #5827.

### LINKED ISSUE #5827: Nim beginner's feedback: "echo type(1)" does not work
I've recently introduced Nim to a number of colleagues (mostly with a C++ and Go background), and I want to share some of the feedback I got.

Surprise: you cannot print the type of an expression
```
echo 1
echo type(1)    # does not compile
```

This seems easy to fix with the following patch (not sure if that is correct), but I wonder why something like this not enabled by default given that the `type()` proc is exported from system.nim.
```
diff --git a/lib/system.nim b/lib/system.nim
index 9b41253..ce41373 100644
--- a/lib/system.nim
+++ b/lib/system.nim
@@ -168,6 +168,8 @@ proc `type`*(x: untyped): typeDesc {.magic: "TypeOf", noSideEffect, compileTime.
   ## Builtin 'type' operator for accessing the type of an expression.
   ## Cannot be overloaded.
   discard
+from typetraits import nil
+proc `$`*(t: typedesc): string = typetraits.name(t)
 
 proc `not` *(x: bool): bool {.magic: "Not", noSideEffect.}
   ## Boolean not; returns true iff ``x == false``.
```

### PR DESCRIPTION
**Pull request by dom96**

### CODE REVIEW

**Comment by Araq** [comment]
I don't want it in system.nim. Nor do I want people to ``echo`` types.

**Comment by bluenote10** [comment]
Why is that so bad? Imho easily echoing types comes in handy

- to-string conversion of generics
- during debugging/development 
- in some meta programming use case e.g. when including type information in an error message

My most basic use case looks like this:

```nimrod
import typetraits

type
  Generic[T] = object
    a, b: T

proc `$`*[T](g: Generic[T]): string =
  # A meaningful string representation should encode T in some way.
  # I often want something like this for clarity:
  "Generic[" & $T & "](" & $g.a & ", " & $g.b & ")"

let g = Generic[int](a: 1, b: 2)
echo g # Generic[int](1, 2)
```

I just went through a bunch of my typetraits imports and they all where because of this.

**Comment by kaushalmodi** [comment]
I also wondered why the basic `foo.type.name` needed importing a separate module. 

For folks like me who like introspection of the code while developing (habit from Emacs lisp), this is a must have.  

Knowing the type of any variable is a requirement for me to help understand Nim better. I now end up importing typetraits everywhere in my notes: https://scripter.co/notes/nim/.

**Comment by Yardanico** [comment]
@kaushalmodi why do you use "intToStr" instead of "parseInt" in your notes?

**Comment by kaushalmodi** [comment]
@Yardanico I use both.. as they are are opposite use cases.. `intToStr` is for int to string conversion and `parseInt` is for string to int conversion.

Though I don't know if `intToStr` is needed at all because `$` works just fine.. if `foo` contains `42` (an int), you can do just `$foo` instead of importing `strutils` and then doing `intToStr(foo)`.

**Comment by Yardanico** [comment]
@kaushalmodi I think intToStr is needed when you need to add leading zeroes to your number

**Comment by krux02** [comment]
@Araq  can you elaborate on what is so bad for this to be in system.nim? I would vote to have this in system.nim. Often when I debug things, I do echo debugging, and it is just nice when it just works on types, too. Currently I always have to look up for that package that I have to import so that echo on the type just works, and that is bothering me.

**Comment by kaushalmodi** [comment]
@Yardanico 

> I think intToStr is needed when you need to add leading zeroes to your number

Let's continue this off-topic discussion here: https://forum.nim-lang.org/t/3874.

**Comment by narimiran** [comment]
Based on the reactions: everybody is for this practical change, Araq is against it.

What is it gonna be? Merge (after fixing the version number in the comments) or `won't fix` and close?

**Comment by kaushalmodi** [comment]
Fails because of:

```
[1m[31mFAIL: [36mtbindtypedesc.nim C[0m
[1m[36mTest "tests/metatype/tbindtypedesc.nim" in category "metatype"[0m
[1m[31mFailure: reNimcCrash[0m
[33mExpected:[0m
```

File: lib/pure/typetraits.nim [hunk 1]
```diff
@@ -10,17 +10,18 @@
 ## This module defines compile-time reflection procs for
 ## working with types
 
-proc name*(t: typedesc): string {.magic: "TypeTrait".}
+proc name*(t: typedesc): string =
   ## Returns the name of the given type.
   ##
+  ## As of Nim v0.18.0 this is just an alias for `$` which is defined in
```
Comment (narimiran) [inline lib/pure/typetraits.nim]: Before merging, change it to 0.19.2 or 0.20, depending on when this will be introduced.

**Comment by Araq** [comment]
Makes ``tests/metatype/tbindtypedesc.nim`` fail.

**Comment by timotheecour** [comment]
fixed test failures, addressed comments and revived in https://github.com/nim-lang/Nim/pull/10071

**Pull request closed without merging**